### PR TITLE
Update docs/transform.ipynb based on new transformer framework.

### PR DESCRIPTION
**Pull request by tanujkhattar**
Fixes https://github.com/quantumlib/Cirq/issues/4960

This notebook provides a high level overview of all the newly added features to Cirq's transformer infrastructure. We can discuss whether we need more details docs on each of the specific topics -- I think a high level overview with links to reference pages of individual methods (i.e. the current PR) should be enough to close https://github.com/quantumlib/Cirq/issues/4960, but I'm curious to hear what others think.

### CODE REVIEW

File: docs/transform.ipynb [hunk 1]
```diff
@@ -99,15 +137,271 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "OLJrXdyTCGev"
+   },
+   "source": [
+    "## `cirq.TransformerContext` to store common configurable options. \n",
+    "You might be wondering what's the benefit of adding the `@cirq.transformer` decorator on the methods / classes as defined above? Does it add any extra functionality to the decorated objects at all? \n",
+    "\n",
+    "Yes! The transformer decorator inspects the `cirq.TransformerContext` argument and automatically appends useful functionality like support for automated logging and recursively running the transformer on nested sub-circutis. \n"
```
Comment (MichaelBroughton) [inline docs/transform.ipynb:178]: Nit: generally philosphy is to instruct the reader and not sell/editorialize where possible.

File: docs/transform.ipynb [hunk 2]
```diff
@@ -99,15 +137,271 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "UvFdvxQ6MZD9"
+   },
+   "source": [
+    "### Built-in support for `deep=True`.\n",
+    "Notice that the custom transformers used above only modified the top-level circuit and had no impact on sub-circuits wrapped inside `cirq.CircuitOperation`.\n",
+    "\n",
```
Comment (MichaelBroughton) [inline docs/transform.ipynb:217]: Maybe add a code cell to further clarify this difference via some output comparison ?

File: docs/transform.ipynb [hunk 3]
```diff
@@ -116,61 +410,90 @@
...
+    "*   **cirq.drop_empty_moments** / **cirq.drop_negligible_operations**:  Removes moments that are empty or operations that have very small effects, respectively.\n",
+    "*   **cirq.eject_phased_paulis**: Pushes X, Y, and PhasedX gates towards the end of the circuit, potentially absorbing Z gates and modifying gates along the way.\n",
+    "*   **cirq.eject_z**:  Pushes Z gates towards the end of the circuit, potentially adjusting phases of gates that they pass through.\n",
+    "*   **cirq.expand_composite**:  Uses `cirq.decompose` to expand composite gates.\n",
+    "*   **cirq.merge_k_qubit_unitaries**: Replaces connected components of unitary operations, acting on <= k qubits, with op-tree given by `rewriter(circuit_op)`.\n",
+    "*   **cirq.optimize_for_target_gateset**: Attempts to convert a circuit into and equivalent circuit using only gates from a given target gateset.\n",
+    "*   **cirq.stratified_circuit:**: Repacks the circuit to ensure that moments only contain operations from the same category.\n",
+    "*   **cirq.synchronize_terminal_measurements**:  Moves all terminal measurements in a circuit to the final moment, if possible."
```
Comment (MichaelBroughton) [inline docs/transform.ipynb:426]: Personally I would delete the laundry list here and just link out to the `cirq.transformers` module so a motivated user could sift through the latest features there. Plus as maintainers we don't have to remember to keep this up to date.

File: docs/transform.ipynb [hunk 4]
```diff
@@ -116,61 +410,90 @@
...
+    "* A bundle of pre/post transformers + an analytical decomposition method that can be used to rewrite an optimized equivalent circuit using gates from this target gateset. \n",
     "\n",
-    "You can create your own optimizers to transform and modify circuits to fit hardware, gate sets, or other requirements.  Optimizers can also be used to generate noise.  See [noise](noise.ipynb) for details.\n",
+    "Cirq has implementations for common target gatesets like `cirq.CZTargetGateset`, `cirq.SqrtIswapTargetGateset` etc.\n",
     "\n",
-    "You can do this by implementing the function `optimize_circuit`.\n",
     "\n",
-    "If your optimizer is a local optimizer and depends primarily on operator being examined, you can alternatively inherit `cirq.PointOptimizer` and implement the function `optimization_at(self, circuit, index, op)` that optimizes a single operation.\n",
+    "**`cirq.optimize_for_target_gateset`** is a transformer which compiles a given circuit for a `cirq.CompilationTargetGateset` gateset via the following steps:\n",
```
Comment (MichaelBroughton) [inline docs/transform.ipynb:445]: It feels like there might be a bit too much information in this text here. Could we cut this down to introduce `optimize_for_target_gateset` maybe demo a built-in gateset and end on some hook to encourage the reader to check out other CompilationTargetGatesets to see how they work ?

File: docs/transform.ipynb [hunk 5]
```diff
@@ -116,61 +410,90 @@
...
+    "id": "-QFPCdW0qS3R"
+   },
+   "outputs": [],
+   "source": [
+    "context = cirq.TransformerContext(logger=cirq.TransformerLogger())\n",
+    "converted_circuit_optimized = cirq.optimize_for_target_gateset(circuit, gateset = cirq.CZTargetGateset(allow_partial_czs=True), context=context)\n",
+    "cirq.testing.assert_circuits_with_terminal_measurements_are_equivalent(circuit, converted_circuit_optimized)\n",
+    "print(f\"Circuit Depth Original: {len(circuit)}\", circuit, sep=\"\\n\")\n",
+    "print(f\"Circuit Depth Converted Custom: {len(converted_circuit)}\", converted_circuit, sep=\"\\n\")\n",
+    "print(f\"Circuit Depth Converted Optimized: {len(converted_circuit_optimized)}\", converted_circuit_optimized, sep=\"\\n\")"
```
Comment (MichaelBroughton) [inline docs/transform.ipynb:470]: Nit: break up into multiple cells and walk through what's happening clearly.

File: docs/transform.ipynb [hunk 6]
```diff
@@ -116,61 +410,90 @@
...
+   "metadata": {
+    "id": "aTNKFcPesNpy"
+   },
+   "source": [
+    "# Summary\n",
+    "Cirq provides a flexible and powerful framework to\n",
+    "* Use built-in transformer primitives and analytical tools to create powerful custom transformers AND \n",
+    "* Easily integrate custom transformers with built-in infrastructure to augment functionality like automated logging, recursive execution on sub-circuits, support for no-compile tags etc.\n",
+    "\n",
+    "Cirq also provides a plethora of built-in transformers which can be composed together into a sensible abstraction, like `cirq.CompilationTargetGateset`, which in-turn can be serialized and can be used as a parameter in larger compilation pipelines and experiment workflows. "
```
Comment (MichaelBroughton) [inline docs/transform.ipynb:496]: Usually we want to end on some kind of "Now you go and build on this for yourself". Could we rework this into something like: "Try using transformers to optimize your circuits for your compilation needs" or "try adapting the above transformers to work with SycamoreGates" or something ?

**Review by MichaelBroughton — COMMENTED** [review]
Good first pass. Left a few high level comments, my biggest concern is that things are a little information dense. We don't need to explain lots of details in the tutorials, but give a nice walkthrough of how things work from a basic to moderately advanced use case and then springboard people into the right parts of the API docs if they want to become experts. IMO I think this would read a lot better if we moved the `Built in transformers in Cirq` section up a little higher and simplified the introduction so that users could learn to be comfortable with the built in transformers, then they could gradually upgrade to a standalone custom function, standalone custom class and then CompilationTargetGateset and then done. Right now it seems a little out of order and a little information dense.

**Comment by tanujkhattar** [comment]
@MichaelBroughton Restructured the entire notebook as per the feedback. We now introduce built-in transformers first, show examples of how to use transformer context and then move on to details about the API / decorator and give examples on how users can create their own transformers. 

Please take another look.

File: docs/transform.ipynb [hunk 7]
```diff
@@ -75,20 +75,58 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "YFrGiPXU2fOG"
+   },
+   "outputs": [],
+   "source": [
+    "from typing import Optional"
+   ]
+  },
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply docs/transform.ipynb:96]: Removed the type hints.

File: docs/transform.ipynb [hunk 8]
```diff
@@ -75,20 +75,58 @@
...
+    "Cirq provides many out-of-the-box transformers which can be used as individual compilation passes and also provides a general framework for users to create their own transformers using powerful primitives and bundle a bunch of transformers together to enable compiling circuits for specific targets. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "jyNdxrkI4CYX"
+   },
+   "source": [
+    "# Transformer API"
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply docs/transform.ipynb:122]: Restructured. PTAL.

File: docs/transform.ipynb [hunk 9]
```diff
@@ -99,15 +137,271 @@
...
    "outputs": [],
    "source": [
-    "import cirq\n",
-    "c=cirq.Circuit()\n",
-    "c.append(cirq.Moment([]))\n",
-    "c.append(cirq.Moment([cirq.X(cirq.GridQubit(1,1))]))\n",
-    "c.append(cirq.Moment([]))\n",
-    "print(f'Before optimization, Circuit has {len(c)} moments')\n",
+    "@cirq.transformer\n",
+    "def reverse_circuit(circuit: 'cirq.AbstractCircuit', *, context: Optional[cirq.TransformerContext] = None) -> 'cirq.AbstractCircuit':\n",
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply docs/transform.ipynb:771]: removed type hints.

File: docs/transform.ipynb (hunk 3 above)
Comment (tanujkhattar) [reply docs/transform.ipynb]: My motivation to keep this list as it is is that `cirq.transformers` reference page on the docs site has too many functions (eg: all functions from transformers.analytical_decompositions, transformers.heuristic_decompositions etc.) and it's hard to look at the reference page and guess which of these are transformers. 

Can we modify the reference page to have a classification of functions into transformers / transformer primitives / analytical decompositions etc?

Till we don't have that, I think it's worth keeping this list here.

File: docs/transform.ipynb (hunk 4 above)
Comment (tanujkhattar) [reply docs/transform.ipynb]: Reduced the information content and restructured. PTAL.

File: docs/transform.ipynb (hunk 6 above)
Comment (tanujkhattar) [reply docs/transform.ipynb]: Added an ending line with link to api reference docs.

File: docs/transform.ipynb (hunk 5 above)
Comment (tanujkhattar) [reply docs/transform.ipynb]: done

File: docs/transform.ipynb (hunk 2 above)
Comment (tanujkhattar) [reply docs/transform.ipynb]: Restructured, this is moot now.

File: docs/transform.ipynb [hunk 10]
```diff
@@ -99,15 +137,271 @@
...
+    "    transformed_circuit = circuit.unfreeze(copy=True)\n",
+    "    transformed_circuit.batch_replace(batch_replace)\n",
+    "    return transformed_circuit\n",
+    "\n",
+    "q = cirq.NamedQubit(\"q\")\n",
+    "circuit = cirq.Circuit(cirq.X(q), cirq.CircuitOperation(cirq.FrozenCircuit(cirq.X(q), cirq.Y(q))), cirq.Z(q))\n",
+    "substitute_gate = SubstituteGate(cirq.X, cirq.S)\n",
+    "print(\"Original Circuit:\", circuit, \"\\n\", sep=\"\\n\")\n",
+    "print(\"Reversed Circuit:\", reverse_circuit(circuit), \"\\n\", sep=\"\\n\")\n",
+    "print(\"Substituted Circuit:\", substitute_gate(circuit), sep=\"\\n\")"
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply docs/transform.ipynb:750]: done

File: docs/transform.ipynb (hunk 1 above)
Comment (tanujkhattar) [reply docs/transform.ipynb]: reworded.

File: docs/transform.ipynb [hunk 11]
```diff
@@ -75,102 +67,992 @@
...
+    "*   **cirq.expand_composite**:  Uses `cirq.decompose` to expand composite gates.\n",
+    "*   **cirq.merge_k_qubit_unitaries**: Replaces connected components of unitary operations, acting on <= k qubits, with op-tree given by `rewriter(circuit_op)`.\n",
+    "*   **cirq.optimize_for_target_gateset**: Attempts to convert a circuit into and equivalent circuit using only gates from a given target gateset.\n",
+    "*   **cirq.stratified_circuit**: Repacks the circuit to ensure that moments only contain operations from the same category.\n",
+    "*   **cirq.synchronize_terminal_measurements**:  Moves all terminal measurements in a circuit to the final moment, if possible.\n",
     "\n",
-    "Optimizers will have a function `optimize_circuit()` that can be used to perform this optimization.  Here is a simple example that removes empty moments:"
+    "\n",
+    "\n",
+    "Below we show a toy example of how to use these transformers to optimize a circuit. "
```
Comment (MichaelBroughton) [inline docs/transform.ipynb:135]: Nit: we show. Second person you might want to do something like "Below, you can implement a transformer pipeline called `optimize_circuit`".

File: docs/transform.ipynb [hunk 12]
```diff
@@ -75,102 +67,992 @@
...
+    "id": "mu83U5hGzNcH"
+   },
+   "source": [
+    "## Inspecting transformer actions\n",
+    "\n",
+    "Every transformer in Cirq accepts a `cirq.TransformerContext` instance, which stores common configurable options useful for all transformers. \n",
+    "\n",
+    "One of the members of transformer context dataclass is `cirq.TransformerLogger` instance. When a logger instance is specified, every cirq transformer logs it's action on the input circuit using the given logger instance. The logs can then be inspected to understand the action of each individual transformer on the circuit. \n",
+    "\n",
+    "Below, we show an example of how to inspect the action of each transformer in the `optimize_circuit` method defined above. "
```
Comment (MichaelBroughton) [inline docs/transform.ipynb:209]: Nit: second person

File: docs/transform.ipynb [hunk 13]
```diff
@@ -75,102 +67,992 @@
...
+      "                                                                                                                            │                                    │                                             │                                          │                                     │                                            │\n",
+      "    2: ─────────────────────────────────────────────────────────────────────────────────────PhXZ(a=0.83,x=0.136,z=-0.83)────iSwap^0.5───PhXZ(a=-1,x=0.364,z=0)───iSwap^0.5───PhXZ(a=-0.573,x=0.467,z=-0.795)───iSwap^0.5───PhXZ(a=0.148,x=0.5,z=-0.148)───iSwap^0.5───PhXZ(a=0.5,x=0.623,z=0)───iSwap^0.5───PhXZ(a=-0.5,x=0.352,z=-0.375)────M───\n",
+      "----------------------------------------\n"
+     ]
+    }
+   ],
+   "source": [
+    "# `cirq.optimize_for_target_gateset` also supports all the features discussed above, using `cirq.TransformerContext`.\n",
+    "# For example, we can compile the circuit for Sqrt-Iswap Target Gateset and inspect action of individual transformers.\n",
```
Comment (MichaelBroughton) [inline docs/transform.ipynb:669]: Nit: second person.

File: docs/transform.ipynb [hunk 14]
```diff
@@ -75,102 +67,992 @@
...
+    "id": "kLm-5LHr0YAd"
+   },
+   "source": [
+    "## Support for no-compile tags. \n",
+    "\n",
+    "Cirq also supports tagging operations with no-compile tags such that these tagged operations are ignored when applying transformations on the circuit. This allows users to gain more fine-grained conrol over the compilation process. \n",
+    "\n",
+    "Any valid tag can be used as a \"no-compile\" tag by adding it to the `tags_to_ignore` field in `cirq.TransformerContext`. When called with a context, cirq transformers will inspect the `context.tags_to_ignore` field and ignore an operation if `op.tags & context.tags_to_ignore` is not empty. \n",
+    "\n",
+    "Below, we show an example of how to use no-compile tags when transforming circuits. #"
```
Comment (MichaelBroughton) [inline docs/transform.ipynb:338]: Nit: second person.

**Review by MichaelBroughton — APPROVED** [review]
Looks awesome. I think this flows really nicely now. A few more nits and then one final step: Generally we want to ship tutorials with no outputs in the cells (and let the docs jobs populate them for us). In our docs infra a tutorial that has pre-computed notebook cells will not get run on the docs infra and instead just be rendered directly to the site. For stability/reliability reasons its often better to leave notebooks without any output if we can.

After that feel free to merge.

**Pull request merged**